File: hrms/hr/report/mr_monthly_overtime_register/mr_monthly_overtime_register.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

# import frappe


from __future__ import unicode_literals
import frappe
from frappe.utils import cstr, cint, getdate, flt
from frappe import msgprint, _
from calendar import monthrange

def execute(filters=None):
	if not filters:
		filters = {}

	conditions, filters = get_conditions(filters)
	columns = get_columns(filters)
	att_map = get_attendance_list(conditions, filters)
	emp_map = get_employee_details(filters.employee_type)

	data = []
	for emp in sorted(att_map):
		emp_det = emp_map.get(emp)
		if not emp_det:
			continue

		row = [emp_det.person_name, emp_det.id_card, emp_det.cost_center]

		total_p = 0.0
		for day in range(filters["total_days_in_month"]):
			for t in ("R","S"):
				status = att_map.get(emp).get(str(day + 1)+str(t), '')

				if flt(status) > 0:
					total_p += flt(status)

				row.append(status)

		row += [total_p]
		data.append(row)

	return columns, data

def get_columns(filters):
	columns = [
		_("Name") + "::140", _("CID")+ "::120", _("Cost Center") + ":Link/Cost Center:100"
	]

	for day in range(filters["total_days_in_month"]):
		columns.append(cstr(day+1)+"R" +"::35")
		columns.append(cstr(day+1)+"S" +"::35")

	columns += [_("Total Hours") + ":Float:100"]
	return columns

def get_attendance_list(conditions, filters):
	attendance_list = frappe.db.sql(f"""select mr_employee as employee, day(date) as day_of_month, cost_center,
					number_of_hours_regular as status, number_of_hours_special as hrs 
				from `tabMuster Roll Overtime Entry` where docstatus = 1 %s order by mr_employee, date""" %
		conditions, filters, as_dict=1)

	att_map = {}
	for d in attendance_list:
		for t in ("R","S"):
			day_of_month = str(d.day_of_month) + str(t)
			att_map.setdefault(d.employee, frappe._dict()).setdefault(day_of_month, "")
			att_map[d.employee][day_of_month] = d.status if t=="R" else d.hrs
			att_map[d.employee]["cost_center"] = d.cost_center
	return att_map

def get_conditions(filters):
	if not (filters.get("month") and filters.get("year")):
		msgprint(_("Please select month and year"), raise_exception=1)

	filters["month"] = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov",
		"Dec"].index(filters.month) + 1

	filters["total_days_in_month"] = monthrange(cint(filters.year), filters.month)[1]

	conditions = " and month(date) = %(month)s and year(date) = %(year)s"
	if filters.get("cost_center"):
		conditions += " and cost_center = \'" + str(filters.cost_center) + "\' "
	return conditions, filters

def get_employee_details(employee_type):
	emp_map = frappe._dict()
	if employee_type == "Muster Roll Employee":
		for d in frappe.db.sql("""select name, person_name, id_card, cost_center
			from `tabMuster Roll Employee`""", as_dict=1):
			emp_map.setdefault(d.name, d)
	else:
		frappe.throw("Select a Employee Type")

	return emp_map

@frappe.whitelist()
def get_years():
	year_list = frappe.db.sql_list("""select distinct YEAR(date) from `tabMuster Roll Overtime Entry` ORDER BY YEAR(date) DESC""")
	if not year_list:
		year_list = [getdate().year]
	
	return "\n".join(str(year) for year in year_list)

File: hrms/hr/report/mr_monthly_overtime_register/mr_monthly_overtime_register.js
// Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["MR Monthly Overtime Register"] = {
	"filters": [
		{
			"fieldname": "month",
			"label": __("Month"),
			"fieldtype": "Select",
			"options": "Jan\nFeb\nMar\nApr\nMay\nJun\nJul\nAug\nSep\nOct\nNov\nDec",
			"default": ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"][frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth()],
		},
		{
			"fieldname": "year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"reqd": 1
	    },
		{
			"fieldname": "employee_type",
			"label": __("Employee Type"),
			"fieldtype": "Select",
			"options": ['Muster Roll Employee'],
			"reqd": 1
		},
		{
			"fieldname": "cost_center",
			"label": __("Cost Center"),
			"fieldtype": "Link",
			"options": "Cost Center",
			"default":"",
			"get_query": function() {
				return {
					'filters': [
						['Cost Center', 'disabled', '!=', '1'],
						['Cost Center', 'is_group', '!=', '1']
					]
				};
			}
		}
	],

	// "onload": function(me) {
	// 	return frappe.call({
	// 		method: "erpnext.projects.report.mr_monthly_overtime_register.mr_monthly_overtime_register.get_years",
	// 		callback: function(r) {
	// 			if (r.message && Array.isArray(r.message)) {
	// 				var year_filter = me.filters_by_name.year;
					
	// 				// Verify that the year_filter exists
	// 				if (year_filter) {
	// 					// Set dropdown options as newline-separated string of years
	// 					year_filter.df.options = r.message.join("\n");
						
	// 					// Set the default year to the first year in the response
	// 					year_filter.df.default = r.message[0];

	// 					// Refresh and set input after setting options and default
	// 					year_filter.refresh();
	// 					year_filter.set_input(year_filter.df.default);
	// 				} else {
	// 					console.error("Year filter not found.");
	// 				}
	// 			} else {
	// 				console.error("Invalid response format:", r);
	// 			}
	// 		}
	// 	});
	// }
	"onload": function (me) {
			return frappe.call({
					method: "hrms.hr.report.mr_monthly_overtime_register.mr_monthly_overtime_register.get_years",
					callback: function (r) {
							var year_filter = me.filters_by_name.year;
							year_filter.df.options = r.message;
							year_filter.df.default = r.message.split("\n")[0];
							year_filter.refresh();
							year_filter.set_input(year_filter.df.default);
					}
			});
	}
};
